url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/client/FaceChooserFrame.java

```diff
@@ -0,0 +1,1007 @@
+            if ((newRow < 0) || (newRow >= rowCount))
+                throw new IllegalArgumentException
+                ("newRow not in range (0 to " + (rowCount-1) + "): " + newRow);
+            if ((newRow >= currentRow) && (newRow < (currentRow + faceRowsHeight)))
+            {
+                return;  // <--- Early return: Already showing ---
+            }
+
+            boolean createdRows = false;  // Any objects instantiated? (Need to pack their layout)
+            int numNewRows;    // How many not currently showing?
+            int newCurRow;     // New first-showing row number
+            int newCurOffset;  // new upper-left corner face ID
+
+            if (newRow < currentRow)
+            {
+                numNewRows = currentRow - newRow;
+                newCurRow = newRow;
+            }
+            else
+            {
+                numNewRows = 1 + (newRow - (currentRow + faceRowsHeight));
+                newCurRow = newRow - faceRowsHeight + 1;
+            }
+            newCurOffset = newCurRow * rowFacesWidth + 1;
+            if (numNewRows > faceRowsHeight)
+                numNewRows = faceRowsHeight;
+
+            int r;
+            if ((numNewRows == faceRowsHeight) || (newRow < currentRow))
+            {
+                // Scroll up, or completely replace visible.
+                if (numNewRows == faceRowsHeight)
+                {
+                    // Completely replace current visible face grid.
+                    for (r = faceRowsHeight - 1; r >= 0; --r)
+                    {
+                        visibleFaceGrid[r].setVisible(false);
+                        visibleFaceGrid[r] = null;
+                    }
+                } else {
+                    // newRow < currentRow:
+                    // Remove current bottom, scroll up
+                    for (r = faceRowsHeight - numNewRows; r < faceRowsHeight; ++r)
+                    {
+                        visibleFaceGrid[r].setVisible(false);
+                        visibleFaceGrid[r] = null;
+                    }
+                    for (r = faceRowsHeight - numNewRows - 1; r >= 0; --r)
+                        visibleFaceGrid[r + numNewRows] = visibleFaceGrid[r];
+                }
+
+                // Add newly-visible
+                int nextId = newCurOffset;  // face ID to add
+                int visR = 0;  // Visual row number to add
+                // in this loop, r = row number in faceGrid to add
+                for (r = newRow; r < (newRow + numNewRows); ++r, ++visR)
+                {
+                    if (faceGrid[r] == null)
+                    {
+                        faceGrid[r] = new FaceChooserRow(nextId);
+                        add(faceGrid[r]);
+                        createdRows = true;
+                    }
+                    visibleFaceGrid[visR] = faceGrid[r];
+                    visibleFaceGrid[visR].setVisible(true);
+                    nextId += rowFacesWidth;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 612:622, 640:650.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.